### ISSUE TITLE
Idea: Use `Range` to create (some) Fuzzers

### ISSUE THREAD

**Issue by jonathanjouty** [elm-explorations/test#89]
I'm a fan of [`hedgehog`](https://hackage.haskell.org/package/hedgehog) [1] and I think it gets a lot of things right.

One of those is having a `Range` parameter for all fuzzers (_i.e._ `Gen`s) that generate values with some notion of a range: integers, floats, strings, lists and other collection types.

Thus, there is no `int : Fuzzer Int`.
Only `int : Range Int -> Fuzzer Int`.

In `hedgehog`, internally a `Range` is defined as (translated to Elm):
```elm
type Size
    = Size Int


type Range a
    = Range a (Size -> ( a, a ))
```
Here `Size` is _essentially_ the number of runs.
In `Range a boundsFunction`:
1. `a` is the "origin" value
2. `boundsFunction` shrinks towards this "origin" `a` as `Size` goes down (example defaults are `0` for `Int`s, empty list for lists)

See [2] for a comment about how I understand `Size` being used in `boundsFunction`.

To construct a `Range Int`, you would use:
```elm
linear : Int -> Int -> Range Int
linear lo hi =
    linearFrom lo lo hi


linearFrom : Int -> Int -> Int -> Range Int
linearFrom origin lo hi = 
    Range origin (intBoundsFunction origin lo hi)

intBoundsFunction : Int -> Int -> Int -> (Size -> (Int, Int))
-- Not elaborated on purpose, you can see hedgehog source for how it's done in Haskell
```

IMO this has several advantages:
1. You get shrinking "for free" as part of the `boundsFunction`, it's useful to look at the `hedgehog` source [3] to see how they do this using `linear`, `linearFrom`, `exponential` and `exponentialFrom` constructor functions.
1. You end up with a more flexible and powerful API.
1. It forces users to think about the range of values they want to generate, so that it matches the problem domain instead of assuming defaults. Both in terms of total range, but also the distribution of values.

I understand that a more flexible and powerful API also results in a more "complicated" API, and you might be trying to go with something simple and uncomplicated.

Hence posting this issue to gauge appetite.

* [1] They also have a more general website with links to packages in other languages: https://hedgehog.qa/
* [2] From a quick scan of the code, it appears if `runs >= 99` the full range is generated, otherwise it is scaled down. So basically the range saturates at `Size 99`.
* [3] https://hackage.haskell.org/package/hedgehog-0.6.1/docs/src/Hedgehog.Internal.Range.html#Range

**Comment by jonathanjouty** [comment]
Well-Typed recently wrote an excellent post discussing both approaches in quite some depth:
https://www.well-typed.com/blog/2019/05/integrated-shrinking/
(although a lot does not apply to Elm, as it considers how both approaches work in their Applicative and Monadic contexts)

Footnote (3) in particular touches on `Range`:
> To some degree we can reduce the need for shrinking by trying small counter examples first; both QuickCheck and Hedgehog do this, though Hedgehog’s approach using first-class “ranges” is arguably nicer here. However, this is not sufficient. It is often the case that the probability that a larger test case hits a given bug is disproportionally larger than the probability that a small test case does, and we are therefore more likely to find bugs in larger test cases than smaller ones.

I will need to dig into `elm-test` code again, but it seems to take the `Simplifier` (AKA shrinker) of the form `a -> LazyList a` and bundle it into a rose tree generator (in `ValidFuzzer`) by mapping `simplifyTree` over the generated values. So the plumbing to work with integrated shrinking exists, it's just not used directly.

Basically I think by using `Range`s we could actually remove the need for manually writing simplifiers/shrinkers.

I might try and develop this idea as a separate library to see how it could work with `elm-test`, before seeing how this can be done _inside_ `elm-test`.

**Comment by Janiczek** [comment]
>[!NOTE]
> A historical note: we have reimplemented elm-test in 2.x.x to the approach from Hypothesis: automatic shrinking by having an intermediate representation ~ `List Int` which gets shrunk (without regard for value types "inside") and then new values get _parsed_ from that.
> 
> That gets around the issue both Hedgehog and QuickCheck have with monads and shrinking.

FWIW, this issue talks about `Range`. I don't clearly see benefit to `Range` in our current situation. Maybe some example would help. Is it that you'd be able to control in your fuzzer definition how large the generated values are based on how many Fuzz runs are left to do? (Right now you can change the distribution with `Fuzz.frequency` but it's independent of the fuzz runs...)

**Comment by jonathanjouty** [comment]
Thanks for pinging on this, and great work on the `2.x.x` rewrite, with which most of my points are resolved 🙌🏽

1. Shrinking for free: solved
2. A more flexible API: partly solved
3. Users must think about the range: not solved, and probably fine

2 & 3 detailed below, but first, an answer to your question.

## Question: `Range` for recursive types

> I don't clearly see benefit to `Range` in our current situation. Maybe some example would help. Is it that you'd be able to control in your fuzzer definition how large the generated values are based on how many Fuzz runs are left to do?

I have no example where the "runs left" **must** control the size. I think Hedgehog mainly uses `Size` for recursive types, through `Gen.recursive`. In `2.x.x` a depth param does the same job, and you can fuzz the depth too:

```elm
tree : Fuzzer Tree
tree =
    Fuzz.intRange 0 5
        |> Fuzz.andThen treeOfDepth

treeOfDepth : Int -> Fuzzer Tree
-- skipped impl.
```

The shrinker makes the depth smaller, so this actually gives more control than Hedgehog (if relying solely on `Size`).

## 2. A more flexible API

`intRange`, `floatRange`, `stringOfLengthBetween` and `listOfLengthBetween` cover plain bounds. I think three gaps remain.

### A) Size guarantees
The first gap is collections with a size guarantee. This fuzzer can give a set with fewer than 3 elements, because duplicates collapse:

```elm
Fuzz.listOfLengthBetween 3 10 (Fuzz.intRange 0 5)
    |> Fuzz.map Set.fromList
```

Hedgehog has `Gen.set (Range.linear 3 10) ...`. A `Fuzz.filter` on the size works, but it discards many inputs. Maybe that's acceptable? Having size-bounded `set`, `dict`, and `Array` fuzzers would close the gap?

### B) Shrink target
The second gap is the shrink target. Every numeric fuzzer shrinks towards 0, or towards the bound nearest to 0. Take a fuzzer for years:

```elm
Fuzz.intRange 1900 2100
```

A failure shrinks to 1900. In Hedgehog, `Range.linearFrom 2000 1900 2100` shrinks to 2000, the value that you pick as normal for the domain. Users can build this today if they centre the range on 0 and then shift it:

```elm
year : Fuzzer Int
year =
    Fuzz.intRange -100 100
        |> Fuzz.map (\n -> 2000 + n)
```

This works because `intRange` shrinks towards 0 on both sides of 0. The `intRange` docs do not say that, so this gap is mostly documentation, but also convenience. A note in the docs, and/or a built-in `intRangeFrom`, would close it.

### C) Modular string fuzzers

A neat Hedgehog design is that there are no plain string fuzzers, and since strings are lists of characters, they force you to think about what characters you want to fuzz:
```hs
genValidPassword = Gen.string (Range.linear 0 100) (Gen.choice [Gen.alphaNum, genValidSpecialChar])
```

Having string fuzzers that take a `Char` fuzzer, rather than just having two options (all Unicode or only ASCII), would fix that.

Adding to that, `Fuzz` has no character-class fuzzers like Hedgehog has ([see their character fuzzers here](https://hackage-content.haskell.org/package/hedgehog-1.7/docs/Hedgehog-Gen.html#g:6)). That said, users can do this themselves with `Fuzz.oneOfValues`, so just an inconvenience.

## 3. Users must think about the range: not solved, and probably fine

`Fuzz.int` still exists without bounds. I no longer think that `elm-test` must remove it, because good defaults help new users. `Test.Distribution` also lets users measure the distribution, and that covers part of my original concern. Users cannot choose a distribution shape such as linear or exponential. I have no strong use case for that.
